pkg/controller/backup_controller.go

```diff
@@ -385,51 +385,75 @@ func (c *backupController) runBackup(backup *pkgbackup.Request) error {
-	backupJSON := new(bytes.Buffer)
-	if err := encode.EncodeTo(backup.Backup, "json", backupJSON); err != nil {
-		errs = append(errs, errors.Wrap(err, "error encoding backup"))
-	} else {
-		// Only upload the json and backup tarball if encoding to json succeeded.
-		backupJSONToUpload = backupJSON
-		backupFileToUpload = backupFile
-	}
+	errs = append(errs, persistBackup(backup, backupFile, logFile, backupStore, c.logger)...)
+	errs = append(errs, recordBackupMetrics(backup.Backup, backupFile, c.metrics))
+
+	log.Info("Backup completed")
+
+	return kerrors.NewAggregate(errs)
+}
+
+func recordBackupMetrics(backup *api.Backup, backupFile *os.File, serverMetrics *metrics.ServerMetrics) error {
+	backupScheduleName := backup.GetLabels()["ark-schedule"]
 
 	var backupSizeBytes int64
+	var err error
 	if backupFileStat, err := backupFile.Stat(); err != nil {
-		errs = append(errs, errors.Wrap(err, "error getting file info"))
+		err = errors.Wrap(err, "error getting file info")
 	} else {
 		backupSizeBytes = backupFileStat.Size()
 	}
+	serverMetrics.SetBackupTarballSizeBytesGauge(backupScheduleName, backupSizeBytes)
 
-	if err := gzippedLogFile.Close(); err != nil {
-		c.logger.WithError(err).Error("error closing gzippedLogFile")
-	}
+	backupDuration := backup.Status.CompletionTimestamp.Time.Sub(backup.Status.StartTimestamp.Time)
+	backupDurationSeconds := float64(backupDuration / time.Second)
+	serverMetrics.RegisterBackupDuration(backupScheduleName, backupDurationSeconds)
 
-	if err := backupStore.PutBackup(backup.Name, backupJSONToUpload, backupFileToUpload, logFile); err != nil {
-		errs = append(errs, err)
+	return err
+}
+
+func persistBackup(backup *pkgbackup.Request, backupContents, backupLog *os.File, backupStore persistence.BackupStore, log logrus.FieldLogger) []error {
+	errs := []error{}
+	backupJSON := new(bytes.Buffer)
+
+	if err := encode.EncodeTo(backup.Backup, "json", backupJSON); err != nil {
+		errs = append(errs, errors.Wrap(err, "error encoding backup"))
+
+		// Don't upload the json or backup tarball if encoding to json fails.
+		backupJSON = nil
+		backupContents = nil
 	}
 
-	backupScheduleName := backup.GetLabels()["ark-schedule"]
-	c.metrics.SetBackupTarballSizeBytesGauge(backupScheduleName, backupSizeBytes)
+	volumeSnapshotList := new(api.VolumeSnapshotList)
+	for _, vs := range backup.VolumeSnapshots {
+		volumeSnapshotList.Items = append(volumeSnapshotList.Items, *vs)
+	}
 
-	backupDuration := backup.Status.CompletionTimestamp.Time.Sub(backup.Status.StartTimestamp.Time)
-	backupDurationSeconds := float64(backupDuration / time.Second)
-	c.metrics.RegisterBackupDuration(backupScheduleName, backupDurationSeconds)
+	volumeSnapshotJSON := new(bytes.Buffer)
+	if err := encode.EncodeTo(volumeSnapshotList, "json", volumeSnapshotJSON); err != nil {
+		errs = append(errs, errors.Wrap(err, "error encoding list of volume snapshots"))
+		// TODO what do we want to do if we fail to encode the VolumeSnapshots list to JSON?
```

Comment:
One answer is log the reason why for our internal consumption and attempt to prevent it in the future if there's a way.  We could also maybe create a json with an error message.